gobblin-api/src/main/java/org/apache/gobblin/password/PasswordManager.java

```diff
@@ -64,34 +70,42 @@
   private static final long CACHE_SIZE = 100;
   private static final long CACHE_EXPIRATION_MIN = 10;
   private static final Pattern PASSWORD_PATTERN = Pattern.compile("ENC\\((.*)\\)");
+  private int NUMBER_OF_ENCRYPT_KEYS;
+  private FileSystem fs;
+  private String masterPasswordFile;
+  private boolean useStrongEncryptor;
 
-  private static final LoadingCache<Map.Entry<Optional<String>, Boolean>, PasswordManager> CACHED_INSTANCES =
+  private static final LoadingCache<Map.Entry<State, Boolean>, PasswordManager> CACHED_INSTANCES =
       CacheBuilder.newBuilder().maximumSize(CACHE_SIZE).expireAfterAccess(CACHE_EXPIRATION_MIN, TimeUnit.MINUTES)
-          .build(new CacheLoader<Map.Entry<Optional<String>, Boolean>, PasswordManager>() {
-
+          .build(new CacheLoader<Map.Entry<State, Boolean>, PasswordManager>() {
             @Override
-            public PasswordManager load(Map.Entry<Optional<String>, Boolean> cacheKey) {
+            public PasswordManager load(Map.Entry<State, Boolean> cacheKey) {
               return new PasswordManager(cacheKey.getKey(), cacheKey.getValue());
             }
           });
 
-  private Optional<TextEncryptor> encryptor;
-
-  private PasswordManager(Optional<String> masterPassword, boolean useStrongEncryptor) {
-    if (masterPassword.isPresent()) {
-      this.encryptor = useStrongEncryptor ? Optional.of((TextEncryptor) new StrongTextEncryptor())
-          : Optional.of((TextEncryptor) new BasicTextEncryptor());
-      try {
+  private Optional<TextEncryptor> encryptor = Optional.absent();
 
-        // setPassword() needs to be called via reflection since the TextEncryptor interface doesn't have this method.
-        this.encryptor.get().getClass().getMethod("setPassword", String.class).invoke(this.encryptor.get(),
-            masterPassword.get());
-      } catch (Exception e) {
-        LOG.error("Failed to set master password for encryptor", e);
-        this.encryptor = Optional.absent();
+  private PasswordManager(State state, boolean useStrongEncryptor) {
+    try {
+      this.NUMBER_OF_ENCRYPT_KEYS = state.getPropAsInt(ConfigurationKeys.NUMBER_OF_ENCRYPT_KEYS, ConfigurationKeys.DEFAULT_NUMBER_OF_MASTER_PASSWORDS);
+      this.useStrongEncryptor = useStrongEncryptor;
+      if (state.contains(ConfigurationKeys.ENCRYPT_KEY_LOC)) {
+        this.masterPasswordFile = state.getProp(ConfigurationKeys.ENCRYPT_KEY_LOC);
+        if (state.contains(ConfigurationKeys.ENCRYPT_KEY_FS_URI)) {
+          this.fs = FileSystem.get(URI.create(state.getProp(ConfigurationKeys.ENCRYPT_KEY_FS_URI)), new Configuration());
+        } else {
+          this.fs = new Path(masterPasswordFile).getFileSystem(new Configuration());
+        }
+        if (!this.fs.exists(new Path(masterPasswordFile)) || fs.getFileStatus(new Path(masterPasswordFile)).isDirectory()) {
```

Comment:
Find all available keys and cache them here by name to avoid creating them on every decrypt.

Reply:
done.